API/Helpers/PdfMetadataExtractor.cs

```diff
@@ -0,0 +1,1605 @@
+        }
+
+        PushMetadataRef(meta);
+    }
+
+    private void PushMetadataRef(MetadataRef meta)
+    {
+        if (metadataRef.Count > 0)
+        {
+            if (meta.root == metadataRef.Peek().root)
+            {
+                meta.root = -1;
+            }
+
+            if (meta.info == metadataRef.Peek().info)
+            {
+                meta.info = -1;
+            }
+        }
+
+        if (meta.root != -1 || meta.info != -1)
+        {
+            metadataRef.Push(meta);
+        }
+    }
+
+    private void ReadTrailerDictionary()
+    {
+        // Read trailer directory
+        long prev = -1;
+        long xrefStm = -1;
+
+        MetadataRef meta = new(-1, -1);
+
+        ParseDictionary(delegate(string key, PdfLexer.Token value)
+        {
+            switch (key)
+            {
+                case "Root":
+                    if (value.type != PdfLexer.TokenType.ObjectRef)
+                    {
+                        throw new Exception("Expected object reference after /Root");
+                    }
+
+                    meta.root = (long)value.value;
+
+                    return true;
+                case "Prev":
+                    if (value.type != PdfLexer.TokenType.Int)
+                    {
+                        throw new Exception("Expected offset after /Prev");
+                    }
+
+                    prev = (long)value.value;
+
+                    return true;
+                case "Info":
+                    if (value.type != PdfLexer.TokenType.ObjectRef)
+                    {
+                        throw new Exception("Expected object reference after /Info");
+                    }
+
+                    meta.info = (long)value.value;
+
+                    return true;
+                case "XRefStm":
+                    // Prefer encoded xref stream over xref table
+                    if (value.type != PdfLexer.TokenType.Int)
+                    {
+                        throw new Exception("Expected offset after /XRefStm");
+                    }
+
+                    xrefStm = (long)value.value;
+
+                    return true;
+
+                case "Encrypt":
+                    throw new Exception("Encryption not supported");
+
+                default:
+                    return false;
+            }
+        });
+
+        PushMetadataRef(meta);
+
+        if (xrefStm != -1)
+        {
+            ReadXRefAndTrailer(xrefStm);
+        }
+
+        if (prev != -1)
+        {
+            ReadXRefAndTrailer(prev);
+        }
+    }
+
+    private void ReadMetadata(string filename)
+    {
+        // We read potential metadata sources in backwards historical order, so
+        // we can overwrite to our hearts content
+        while (metadataRef.Count > 0)
+        {
+            var meta = metadataRef.Pop();
+
+            _logger.LogDebug("DocumentCatalog for {Path}: {Root}, Info: {Info}", filename, meta.root, meta.info);
```

Comment:
Make all your Debugs into Trace. Debug is normal for users to be on with Kavita.